### PR TITLE
Add UI for viewing and editing charm configuration

### PR DESCRIPTION
**Pull request by hatched**
## Done

Add a UI for viewing and editing charm configuration.

## QA

Note: each PR from the feature branch has already been reviewed and qa'd but this is for all the marbles!

- View the demo link.
- Navigate to a model and view the configuration for an application.
- See if you can break it.
- Try it out on mobile/mobile views

## Screenshots

Screen Shot 2021-01-25 at 12 54 35 PM

### CODE REVIEW

**Comment by long-chung** [comment]
When >1 values have changed reset default CTA doesn't work

**Comment by anthonydillon** [comment]
QA looks great. Thanks. 

One thing, when applying a change it fails and resets the value back without a message so seems broken. Could we provide some feedback to the user if unsuccessful?

**Comment by hatched** [comment]
@long-chung sorry about that, I've addressed that issue now. If you wouldn't mind taking another look.

**Comment by hatched** [comment]
@anthonydillon yep I agree, an issue was created for that exact issue :) It needs design input though so it's currently blocked from the implementation side.

File: src/components/panels/ConfigPanel/BooleanConfig.tsx [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        "config-input--focused": inputFocused,
+        "config-input--changed": inputChanged,
+      })}
+      onClick={() => setSelectedConfig(config)}
+    >
+      <h5
+        className="u-float-left"
+        onClick={() => setShowDescription(!showDescription)}
+        onKeyPress={() => setShowDescription(!showDescription)}
+        // eslint-disable-next-line
```
Comment (barrymcgee) [inline src/components/panels/ConfigPanel/BooleanConfig.tsx:78]: Can you comment why this is disabled?

**Comment by long-chung** [comment]
Sorry, didn't realise this incorporates the mobile layout work as well.

- [ ] Reset all values CTA needs right margin. It is currently on the edge of the viewport.
Screenshot 2021-01-26 at 16 04 14

**Comment by hatched** [comment]
Thanks @long-chung good find. It's now been resolved.

File: src/components/panels/ConfigPanel/ConfigPanel.tsx [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+                  "p-button--positive config-panel__save-button",
+                  {
+                    "is-active": savingConfig,
+                  }
+                )}
+                onClick={handleSubmit}
+                disabled={!enableSave}
+              >
+                {!savingConfig ? (
+                  "Save & apply"
```
Comment (barrymcgee) [inline src/components/panels/ConfigPanel/ConfigPanel.tsx:205]:
```suggestion
                  "Save &amp; apply"
```

File: src/components/panels/ConfigPanel/ConfigPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    }
+  });
+
+  return elements;
+}
+
+function NoConfigMessage() {
+  return (
+    <div className="config-panel__message">
+      <img src={boxImage} alt="3d box" className="config-panel--center-img" />
```
Comment (barrymcgee) [inline src/components/panels/ConfigPanel/ConfigPanel.tsx:294]:
```suggestion
      <img src={boxImage} alt="" className="config-panel--center-img" />
```

File: src/components/panels/ConfigPanel/ConfigPanel.tsx [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+    </div>
+  );
+}
+
+function NoDescriptionMessage() {
+  return (
+    <div className="config-panel__message">
+      <img
+        src={bulbImage}
+        alt="lightbulb"
```
Comment (barrymcgee) [inline src/components/panels/ConfigPanel/ConfigPanel.tsx:305]:
```suggestion
        alt=""
```

File: src/components/panels/ConfigPanel/TextAreaConfig.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        "config-input--changed": inputChanged,
+      })}
+      data-config-name={config.name}
+      onClick={() => setSelectedConfig(config)}
+    >
+      <h5
+        className="u-float-left"
+        onClick={() => setShowDescription(!showDescription)}
+        onKeyPress={() => setShowDescription(!showDescription)}
+        // eslint-disable-next-line
```
Comment (barrymcgee) [inline src/components/panels/ConfigPanel/TextAreaConfig.tsx:69]: Remove the click from the h5

File: src/components/panels/ConfigPanel/_config-panel.scss [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  &--focused {
+    background-color: #ddedfe;
+  }
+
+  &--changed {
+    background-color: $color-light;
+  }
+
+  &.config-input--focused.config-input--changed {
+    background-color: #ddedfe;
```
Comment (barrymcgee) [inline src/components/panels/ConfigPanel/_config-panel.scss:152]: This should be moved to `settings.scss`

File: src/components/panels/ConfigPanel/_config-panel.scss [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  @media screen and (min-width: $breakpoint-large) {
+    textarea {
+      margin-bottom: 1.2rem !important;
+    }
+  }
+
+  textarea {
+    border: 1px solid #999;
```
Comment (barrymcgee) [inline src/components/panels/ConfigPanel/_config-panel.scss:205]: Also move to `settings.scss`

**Review by barrymcgee — APPROVED** [review]

**Comment by ziheliu214** [comment]
Hi from UX perspective:

1) I assume the fields with default values are compulsory, the rest is optional? Cannot see any signs from the panel itself when first landed.
Screenshot 2021-01-26 at 17 37 15
2) After entering values of the first 2 boxes, the bg of the fields are grey - is it on purpose?
Screenshot 2021-01-26 at 17 43 16
3) same screenshot from above, I can save with the wrong type of data or without entering. Is there a way we prompt the right data format?
4) After saving with wrong data input, it would return ALL fields to the default values without error/explaination messages.See below before & after.
Screenshot 2021-01-26 at 17 47 49
Screenshot 2021-01-26 at 17 47 37

**Comment by hatched** [comment]
@ziheliu214 
1. There is no concept of a compulsary field, those fields with values are just what the value is set to.
2. Yes the grey background indicates that the value has changed locally but has not yet been saved.
3. Form validation and error messages haven't yet been implemented. Issues exist to implement these features.

**Comment by hatched** [comment]
Thanks everyone!

**Pull request merged**